TiFlash PageStorage V3 crash after restart

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
Enable PageStorage V3  and encryption, run CH-benCHmark 1500 warehouses, after two OOM, TiFlash throw exception when restart.
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
TiFlash successfully restart.
### 3. What did you see instead (Required)
TiFlash throw exception when restart.

[2022/05/07 17:13:04.294 +08:00] [ERROR] [<unknown>] ["Application:DB::Exception: Something wrong while reading log file"] [thread_id=1]
[2022/05/07 17:15:19.203 +08:00] [ERROR] [Exception.cpp:85] ["bool DB::Context::initializeGlobalStoragePoolIfNeed(const DB::PathPool &, bool):Code: 0, e.displayText() = DB::Exception: [header_off=11] [payload_length=12662] [checksum=0x3A70CB628F571380] [actual_checksum=0xDCC2066722BB4097] [file=/data2/chentongli/tidb-data/tiflash-19111/page/log/wal/log_5_1], e.what() = DB::Exception, Stack trace:\n\n\n       0x4446e8e\tStackTrace::StackTrace() [tiflash+71593614]\n                \tdbms/src/Common/StackTrace.cpp:23\n       0x44440d2\tDB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+71581906]\n                \tdbms/src/Common/Exception.h:41\n       0xdae76ae\tDB::PS::V3::LogReader::readPhysicalRecord(std::__1::basic_string_view<char, std::__1::char_traits<char> >*, unsigned long*) [tiflash+229537454]\n                \tdbms/src/Storages/Page/V3/LogFile/LogReader.cpp:414\n       0xdae6601\tDB::PS::V3::LogReader::readRecord() [tiflash+229533185]\n                \tdbms/src/Storages/Page/V3/LogFile/LogReader.cpp:122\n       0xdadceef\tDB::PS::V3::WALStoreReader::next() [tiflash+229494511]\n                \tdbms/src/Storages/Page/V3/WAL/WALReader.cpp:179\n       0xdaa76bf\tDB::PS::V3::PageDirectoryFactory::loadFromDisk(std::__1::unique_ptr<DB::PS::V3::PageDirectory, std::__1::default_delete<DB::PS::V3::PageDirectory> > const&, std::__1::shared_ptr<DB::PS::V3::WALStoreReader>&&) [tiflash+229275327]\n                \tdbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:175\n       0xdaa73c8\tDB::PS::V3::PageDirectoryFactory::create(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::shared_ptr<DB::FileProvider>&, std::__1::shared_ptr<DB::PSDiskDelegator>&, DB::PS::V3::WALStore::Config) [tiflash+229274568]\n                \tdbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:29\n       0xda5c248\tDB::PS::V3::PageStorageImpl::restore() [tiflash+228966984]\n                \tdbms/src/Storages/Page/V3/PageStorageImpl.cpp:55\n       0xd625578\tDB::DM::GlobalStoragePool::restore() [tiflash+224548216]\n                \tdbms/src/Storages/DeltaMerge/StoragePool.cpp:116\n       0xcad9212\tDB::Context::initializeGlobalStoragePoolIfNeed(DB::PathPool const&, bool) [tiflash+212701714]\n                \tdbms/src/Interpreters/Context.cpp:1595\n       0x448b93b\tDB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) [tiflash+71874875]\n                \tdbms/src/Server/Server.cpp:1120\n       0xe380f15\tPoco::Util::Application::run() [tiflash+238554901]\n                \tcontrib/poco/Util/src/Application.cpp:335\n       0xe39bb25\tPoco::Util::ServerApplication::run() [tiflash+238664485]\n                \tcontrib/poco/Util/src/ServerApplication.cpp:94\n       0xe39bd03\tPoco::Util::ServerApplication::run(int, char**) [tiflash+238664963]\n                \tcontrib/poco/Util/src/ServerApplication.cpp:618\n       0x4495228\tmainEntryClickHouseServer(int, char**) [tiflash+71914024]\n                \tdbms/src/Server/Server.cpp:1545\n       0x4414f9f\tmain [tiflash+71389087]\n                \tdbms/src/Server/main.cpp:167\n  0x7f4c0a9d7555\t__libc_start_main [libc.so.6+140629]\n       0x4414969\t<unknown symbol> [tiflash+71387497]\n                \t/root/yuzhao/glibc-2.17/csu/../sysdeps/x86_64/start.S:123"] [thread_id=1]
[2022/05/07 17:15:19.300 +08:00] [ERROR] [<unknown>] ["Application:DB::Exception: [header_off=11] [payload_length=12662] [checksum=0x3A70CB628F571380] [actual_checksum=0xDCC2066722BB4097] [file=/data2/chentongli/tidb-data/tiflash-19111/page/log/wal/log_5_1]"] [thread_id=1]
[2022/05/07 17:31:02.816 +08:00] [ERROR] [Exception.cpp:85] ["bool DB::Context::initializeGlobalStoragePoolIfNeed(const DB::PathPool &, bool):Code: 0, e.displayText() = DB::Exception: [header_off=11] [payload_length=1657] [checksum=0x9BE2C0FA71209210] [actual_checksum=0x9D0DD04ACBC4BB34] [file=/data2/chentongli/tidb-data/tiflash-19111/page/meta/wal/log_5_1], e.what() = DB::Exception, Stack trace:\n\n\n       0x4446e8e\tStackTrace::StackTrace() [tiflash+71593614]\n                \tdbms/src/Common/StackTrace.cpp:23\n       0x44440d2\tDB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) [tiflash+71581906]\n                \tdbms/src/Common/Exception.h:41\n       0xdae76ae\tDB::PS::V3::LogReader::readPhysicalRecord(std::__1::basic_string_view<char, std::__1::char_traits<char> >*, unsigned long*) [tiflash+229537454]\n                \tdbms/src/Storages/Page/V3/LogFile/LogReader.cpp:414\n       0xdae6601\tDB::PS::V3::LogReader::readRecord() [tiflash+229533185]\n                \tdbms/src/Storages/Page/V3/LogFile/LogReader.cpp:122\n       0xdadceef\tDB::PS::V3::WALStoreReader::next() [tiflash+229494511]\n                \tdbms/src/Storages/Page/V3/WAL/WALReader.cpp:179\n       0xdaa76bf\tDB::PS::V3::PageDirectoryFactory::loadFromDisk(std::__1::unique_ptr<DB::PS::V3::PageDirectory, std::__1::default_delete<DB::PS::V3::PageDirectory> > const&, std::__1::shared_ptr<DB::PS::V3::WALStoreReader>&&) [tiflash+229275327]\n                \tdbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:175\n       0xdaa73c8\tDB::PS::V3::PageDirectoryFactory::create(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::shared_ptr<DB::FileProvider>&, std::__1::shared_ptr<DB::PSDiskDelegator>&, DB::PS::V3::WALStore::Config) [tiflash+229274568]\n                \tdbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:29\n       0xda5c248\tDB::PS::V3::PageStorageImpl::restore() [tiflash+228966984]\n                \tdbms/src/Storages/Page/V3/PageStorageImpl.cpp:55\n       0xd62557c\tDB::DM::GlobalStoragePool::restore() [tiflash+224548220]\n                \tdbms/src/Storages/DeltaMerge/StoragePool.cpp:116\n       0xcad9212\tDB::Context::initializeGlobalStoragePoolIfNeed(DB::PathPool const&, bool) [tiflash+212701714]\n                \tdbms/src/Interpreters/Context.cpp:1595\n       0x448b93b\tDB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) [tiflash+71874875]\n                \tdbms/src/Server/Server.cpp:1120\n       0xe380f15\tPoco::Util::Application::run() [tiflash+238554901]\n                \tcontrib/poco/Util/src/Application.cpp:335\n       0xe39bb25\tPoco::Util::ServerApplication::run() [tiflash+238664485]\n                \tcontrib/poco/Util/src/ServerApplication.cpp:94\n       0xe39bd03\tPoco::Util::ServerApplication::run(int, char**) [tiflash+238664963]\n                \tcontrib/poco/Util/src/ServerApplication.cpp:618\n       0x4495228\tmainEntryClickHouseServer(int, char**) [tiflash+71914024]\n                \tdbms/src/Server/Server.cpp:1545\n       0x4414f9f\tmain [tiflash+71389087]\n                \tdbms/src/Server/main.cpp:167\n  0x7f6343fab555\t__libc_start_main [libc.so.6+140629]\n       0x4414969\t<unknown symbol> [tiflash+71387497]\n                \t/root/yuzhao/glibc-2.17/csu/../sysdeps/x86_64/start.S:123"] [thread_id=1]
[2022/05/07 17:31:02.989 +08:00] [ERROR] [<unknown>] ["Application:DB::Exception: [header_off=11] [payload_length=1657] [checksum=0x9BE2C0FA71209210] [actual_checksum=0x9D0DD04ACBC4BB34] [file=/data2/chentongli/tidb-data/tiflash-19111/page/meta/wal/log_5_1]"] [thread_id=1]
### 4. What is your TiFlash version? (Required)
04da47f1e848012cb465705894c8ecdd3d81a40b

--- a/dbms/src/Storages/Page/V3/LogFile/LogReader.cpp
+++ b/dbms/src/Storages/Page/V3/LogFile/LogReader.cpp
@@ -470,7 +470,7 @@ UInt8 LogReader::readMore(size_t * drop_size)
 
 void LogReader::reportCorruption(size_t bytes, const String & reason)
 {
-    reportDrop(bytes, "Corruption: " + reason);
+    reportDrop(bytes, fmt::format("Corruption: {} [offset={}] [file={}]", reason, file->getPositionInFile(), file->getFileName()));
 }
 
 void LogReader::reportDrop(size_t bytes, const String & reason)

--- a/dbms/src/Storages/Page/V3/WAL/WALReader.h
+++ b/dbms/src/Storages/Page/V3/WAL/WALReader.h
@@ -20,6 +20,11 @@
 
 namespace DB
 {
+namespace ErrorCodes
+{
+extern const int CORRUPTED_DATA;
+}
+
 class FileProvider;
 using FileProviderPtr = std::shared_ptr<FileProvider>;
 
@@ -32,7 +37,7 @@ public:
     {
         error_happened = true;
         // FIXME: store the reason of corruption
-        throw Exception(msg);
+        throw Exception(msg, ErrorCodes::CORRUPTED_DATA);
     }
 
     bool hasError() const

--- a/dbms/src/Storages/Page/V3/WAL/WALReader.h
+++ b/dbms/src/Storages/Page/V3/WAL/WALReader.h
@@ -28,10 +28,11 @@ namespace PS::V3
 class ReportCollector : public LogReader::Reporter
 {
 public:
-    void corruption(size_t /*bytes*/, const String & /*msg*/) override
+    void corruption(size_t /*bytes*/, const String & msg) override
     {
         error_happened = true;
         // FIXME: store the reason of corruption
+        throw Exception(msg);
     }
 
     bool hasError() const

--- a/dbms/src/Storages/Page/V3/WALStore.cpp
+++ b/dbms/src/Storages/Page/V3/WALStore.cpp
@@ -203,13 +203,14 @@ bool WALStore::saveSnapshot(FilesSnapshot && files_snap, PageEntriesEdit && dire
     const auto normal_fullname = log_filename.fullname(LogFileStage::Normal);
 
     LOG_FMT_INFO(logger, "Renaming log file to be normal [fullname={}]", temp_fullname);
-    // Need link encryption info before rename
-    provider->linkEncryptionInfo(temp_fullname, normal_fullname);
 
-    auto f = Poco::File{temp_fullname};
-    f.renameTo(normal_fullname);
-    // Also need delete old encryption info after rename
-    provider->deleteEncryptionInfo(temp_fullname);
+    // Use `renameFile` from FileProvider replace to `renameTo` in Poco::File
+    provider->renameFile(temp_fullname,
+                         EncryptionPath(temp_fullname, ""),
+                         normal_fullname,
+                         EncryptionPath(normal_fullname, ""),
+                         true);
+
     LOG_FMT_INFO(logger, "Rename log file to normal done [fullname={}]", normal_fullname);
 
     // #define ARCHIVE_COMPACTED_LOGS // keep for debug

--- a/dbms/src/Storages/Page/V3/WALStore.cpp
+++ b/dbms/src/Storages/Page/V3/WALStore.cpp
@@ -203,14 +203,15 @@ bool WALStore::saveSnapshot(FilesSnapshot && files_snap, PageEntriesEdit && dire
     const auto normal_fullname = log_filename.fullname(LogFileStage::Normal);
 
     LOG_FMT_INFO(logger, "Renaming log file to be normal [fullname={}]", temp_fullname);
-
-    // Use `renameFile` from FileProvider replace to `renameTo` in Poco::File
-    provider->renameFile(temp_fullname,
-                         EncryptionPath(temp_fullname, ""),
-                         normal_fullname,
-                         EncryptionPath(normal_fullname, ""),
-                         true);
-
+    auto f = Poco::File{temp_fullname};
+    f.renameTo(normal_fullname);
+    // // Use `renameFile` from FileProvider that take good care of encryption path
+    // provider->renameFile(
+    //     temp_fullname,
+    //     EncryptionPath(temp_fullname, ""),
+    //     normal_fullname,
+    //     EncryptionPath(normal_fullname, ""),
+    //     true);
     LOG_FMT_INFO(logger, "Rename log file to normal done [fullname={}]", normal_fullname);
 
     // #define ARCHIVE_COMPACTED_LOGS // keep for debug

--- a/dbms/src/Storages/Page/V3/WALStore.cpp
+++ b/dbms/src/Storages/Page/V3/WALStore.cpp
@@ -201,7 +201,11 @@ bool WALStore::saveSnapshot(FilesSnapshot && files_snap, PageEntriesEdit && dire
     // Rename it to be a normal log file.
     const auto temp_fullname = log_filename.fullname(LogFileStage::Temporary);
     const auto normal_fullname = log_filename.fullname(LogFileStage::Normal);
+
     LOG_FMT_INFO(logger, "Renaming log file to be normal [fullname={}]", temp_fullname);
+    // Need link encryption info before rename
+    provider->linkEncryptionInfo(temp_fullname, normal_fullname);
+
     auto f = Poco::File{temp_fullname};
     f.renameTo(normal_fullname);
     LOG_FMT_INFO(logger, "Rename log file to normal done [fullname={}]", normal_fullname);

--- a/dbms/src/Storages/Page/V3/WALStore.cpp
+++ b/dbms/src/Storages/Page/V3/WALStore.cpp
@@ -208,6 +208,8 @@ bool WALStore::saveSnapshot(FilesSnapshot && files_snap, PageEntriesEdit && dire
 
     auto f = Poco::File{temp_fullname};
     f.renameTo(normal_fullname);
+    // Also need delete old encryption info after rename
+    provider->deleteEncryptionInfo(temp_fullname);
     LOG_FMT_INFO(logger, "Rename log file to normal done [fullname={}]", normal_fullname);
 
     // #define ARCHIVE_COMPACTED_LOGS // keep for debug

--- a/dbms/src/Storages/Page/V3/WALStore.cpp
+++ b/dbms/src/Storages/Page/V3/WALStore.cpp
@@ -203,15 +203,13 @@ bool WALStore::saveSnapshot(FilesSnapshot && files_snap, PageEntriesEdit && dire
     const auto normal_fullname = log_filename.fullname(LogFileStage::Normal);
 
     LOG_FMT_INFO(logger, "Renaming log file to be normal [fullname={}]", temp_fullname);
-    auto f = Poco::File{temp_fullname};
-    f.renameTo(normal_fullname);
-    // // Use `renameFile` from FileProvider that take good care of encryption path
-    // provider->renameFile(
-    //     temp_fullname,
-    //     EncryptionPath(temp_fullname, ""),
-    //     normal_fullname,
-    //     EncryptionPath(normal_fullname, ""),
-    //     true);
+    // Use `renameFile` from FileProvider that take good care of encryption path
+    provider->renameFile(
+        temp_fullname,
+        EncryptionPath(temp_fullname, ""),
+        normal_fullname,
+        EncryptionPath(normal_fullname, ""),
+        true);
     LOG_FMT_INFO(logger, "Rename log file to normal done [fullname={}]", normal_fullname);
 
     // #define ARCHIVE_COMPACTED_LOGS // keep for debug

--- a/dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp
+++ b/dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp
@@ -552,8 +552,7 @@ try
     std::uniform_int_distribution<> d_20(0, 20);
 
     // Stage 2. insert many edits
-    // constexpr size_t num_edits_test = 100000;
-    constexpr size_t num_edits_test = 10;
+    constexpr size_t num_edits_test = 100000;
     PageId page_id = 0;
     std::vector<size_t> size_each_edit;
     size_each_edit.reserve(num_edits_test);

--- a/dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp
+++ b/dbms/src/Storages/Page/V3/tests/gtest_wal_store.cpp
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <Encryption/MockKeyManager.h>
 #include <Poco/Logger.h>
 #include <Storages/Page/V3/LogFile/LogFilename.h>
 #include <Storages/Page/V3/LogFile/LogFormat.h>
@@ -220,6 +221,7 @@ class WALStoreTest
 public:
     WALStoreTest()
         : multi_paths(GetParam())
+        , log(Logger::get("WALStoreTest"))
     {
     }
 
@@ -250,11 +252,11 @@ private:
 protected:
     PSDiskDelegatorPtr delegator;
     WALStore::Config config;
+    LoggerPtr log;
 };
 
 TEST_P(WALStoreTest, FindCheckpointFile)
 {
-    LoggerPtr log = Logger::get("WALLognameTest");
     auto path = getTemporaryPath();
 
     {
@@ -547,10 +549,11 @@ try
     ASSERT_NE(wal, nullptr);
 
     std::mt19937 rd;
-    std::uniform_int_distribution<> d(0, 20);
+    std::uniform_int_distribution<> d_20(0, 20);
 
     // Stage 2. insert many edits
-    constexpr size_t num_edits_test = 100000;
+    // constexpr size_t num_edits_test = 100000;
+    constexpr size_t num_edits_test = 10;
     PageId page_id = 0;
     std::vector<size_t> size_each_edit;
     size_each_edit.reserve(num_edits_test);
@@ -559,7 +562,7 @@ try
     {
         PageEntryV3 entry{.file_id = 2, .size = 1, .tag = 0, .offset = 0x123, .checksum = 0x4567};
         PageEntriesEdit edit;
-        const size_t num_pages_put = d(rd);
+        const size_t num_pages_put = d_20(rd);
         for (size_t p = 0; p < num_pages_put; ++p)
         {
             page_id += 1;
@@ -595,23 +598,45 @@ try
 
     LOG_FMT_INFO(&Poco::Logger::get("WALStoreTest"), "Done test for {} persist pages in {} edits", num_pages_read, num_edits_test);
 
-    // Stage 3. compact logs and verify
-    // wal->compactLogs();
-    // wal.reset();
+    // Test for save snapshot (with encryption)
+    auto enc_key_manager = std::make_shared<MockKeyManager>(/*encryption_enabled_=*/true);
+    auto enc_provider = std::make_shared<FileProvider>(enc_key_manager, true);
+    LogFilenameSet persisted_log_files = WALStoreReader::listAllFiles(delegator, log);
+    WALStore::FilesSnapshot file_snap{.current_writting_log_num = 100, // just a fake value
+                                      .persisted_log_files = persisted_log_files};
 
-    // // After logs compacted, they should be written as one edit.
-    // num_edits_read = 0;
-    // num_pages_read = 0;
-    // wal = WALStore::create(
-    //     [&](PageEntriesEdit && edit) {
-    //         num_pages_read += edit.size();
-    //         EXPECT_EQ(page_id, edit.size()) << fmt::format("at idx={}", num_edits_read);
-    //         num_edits_read += 1;
-    //     },
-    //     provider,
-    //     delegator);
-    // EXPECT_EQ(num_edits_read, 1);
-    // EXPECT_EQ(num_pages_read, page_id);
+    PageEntriesEdit snap_edit;
+    PageEntryV3 entry{.file_id = 2, .size = 1, .tag = 0, .offset = 0x123, .checksum = 0x4567};
+    std::uniform_int_distribution<> d_10000(0, 10000);
+    // just fill in some random entry
+    for (size_t i = 0; i < 70; ++i)
+    {
+        snap_edit.varEntry(d_10000(rd), PageVersionType(345, 22), entry, 1);
+    }
+    std::tie(wal, reader) = WALStore::create(getCurrentTestName(), enc_provider, delegator, config);
+    bool done = wal->saveSnapshot(std::move(file_snap), std::move(snap_edit));
+    ASSERT_TRUE(done);
+    wal.reset();
+    reader.reset();
+
+    // After logs compacted, they should be written as one edit.
+    num_edits_read = 0;
+    num_pages_read = 0;
+    std::tie(wal, reader) = WALStore::create(getCurrentTestName(), enc_provider, delegator, config);
+    while (reader->remained())
+    {
+        auto [ok, edit] = reader->next();
+        if (!ok)
+        {
+            reader->throwIfError();
+            // else it just run to the end of file.
+            break;
+        }
+        num_pages_read += edit.size();
+        num_edits_read += 1;
+    }
+    EXPECT_EQ(num_edits_read, 1);
+    EXPECT_EQ(num_pages_read, 70);
 }
 CATCH
 

